crates/contracts/solidity/Trader.sol

```diff
@@ -13,78 +13,72 @@ contract Trader {
+        _fallback();
+    }
+    receive() external payable {
+        _fallback();
+    }
+    function _fallback() private {
+        bytes memory rdata = TRADER_IMPL.doDelegatecall(msg.data);
+        assembly { return(add(rdata, 32), mload(rdata)) }
+    }
 
-        gasUsed = address(SETTLEMENT).doMeteredCallNoReturn(
-            abi.encodeCall(
-                SETTLEMENT.settle,
-                (tokens, clearingPrices, trades, interactions)
-            )
-        );
+    /// @dev Prepares everything needed by the trader for successfully executing the swap.
+    /// This includes giving the required approval, wrapping the required ETH (if needed)
+    /// and warming the needed storage for sending native ETH to smart contracts.
+    /// @param sellToken - token being sold by the trade
+    /// @param sellAmount - expected amount to be sold according to the quote
+    /// @param nativeToken - ERC20 version of the chain's native token
+    /// @param receiver - address that will receive the bought tokens
+    function prepareSwap(
+        address sellToken,
+        uint256 sellAmount,
+        address nativeToken,
+        address payable receiver
+    ) external {
+        require(!_alreadyCalled, "prepareSwap can only be called once");
+        _alreadyCalled = true;
 
-        for (uint256 i; i < tokens.length; ++i) {
-            traderBalances[i] += IERC20(tokens[i]).balanceOf(address(this)).toInt();
-            settlementBalances[i] += IERC20(tokens[i]).balanceOf(address(SETTLEMENT)).toInt();
+        if (sellToken == nativeToken) {
+            uint256 availableBalance = IERC20(sellToken).balanceOf(address(this));
+            if (availableBalance < sellAmount) {
+                // Simulate wrapping the missing `ETH` so the user doesn't have to spend gas
+                // on that just to get a quote. If they are happy with the quote and want to
+                // create an order they will actually have to do the wrapping, though.
+                (bool sent, ) = nativeToken.call{value: sellAmount - availableBalance}("");
+                require(sent, "'from' address does not have enough (W)ETH balance");
+            }
         }
+
+        uint256 currentAllowance = IERC20(sellToken).allowance(address(this), address(SETTLEMENT.vaultRelayer()));
+        if (currentAllowance < sellAmount) {
+            // Simulate an approval to the settlement contract so the user doesn't have to
+            // spend gas on that just to get a quote. If they are happy with the quote and
+            // want to create an order they will actually have to do the approvals, though.
+            IERC20(sellToken).safeApprove(address(SETTLEMENT.vaultRelayer()), 0);
+            IERC20(sellToken).safeApprove(address(SETTLEMENT.vaultRelayer()), type(uint256).max);
+        }
+
+        (bool sent2, ) = receiver.call{value: 0}("");
+        require(sent2, "warming the storage for sending ETH to smart contract addresses failed");
```

Comment:
Cleaver. I don't think you need to check the `success` here (as the transfer would fail if done as part of a settlement).

A concrete reason for this is that an order can send ERC20 tokens to a contract that doesn't necessarily support receiving Ether, and unless the settlement is transferring Ether to the receiver, we don't want to quote simulation to fail:

```suggestion
        // warming the storage for sending ETH to smart contract addresses
        receiver.call{value: 0}("");
```